src/phillips_dessouky.rs

```diff
@@ -1,80 +1,308 @@
 use pyo3::prelude::*;
 
+use std::collections::{HashMap, HashSet, VecDeque};
+use log::{debug, error, Level};
 use ordered_float::OrderedFloat;
-use pathfinding::directed::edmonds_karp::{
-    SparseCapacity,
-    Edge,
-    edmonds_karp
-};
 
-use std::time::Instant;
-use log::info;
+use crate::lowtime_graph::{LowtimeGraph, LowtimeEdge};
 
 
 #[pyclass]
 pub struct PhillipsDessouky {
-    node_ids: Vec<u32>,
-    source_node_id: u32,
-    sink_node_id: u32,
-    edges_raw: Vec<((u32, u32), f64)>,
+   dag: LowtimeGraph,
+   fp_error: f64,
 }
 
 #[pymethods]
 impl PhillipsDessouky {
     #[new]
     fn new(
+        fp_error: f64,
         node_ids: Vec<u32>,
         source_node_id: u32,
         sink_node_id: u32,
-        edges_raw: Vec<((u32, u32), f64)>,
+        edges_raw: Vec<((u32, u32), (f64, f64, f64, f64))>,
     ) -> PyResult<Self> {
         Ok(PhillipsDessouky {
-            node_ids,
-            source_node_id,
-            sink_node_id,
-            edges_raw,
+            dag: LowtimeGraph::of_python(
+                node_ids,
+                source_node_id,
+                sink_node_id,
+                edges_raw,
+            ),
+            fp_error,
         })
     }
 
-    fn max_flow(&self) -> PyResult<Vec<((u32, u32), f64)>> {
-        let profiling_start = Instant::now();
-        let edges_edmonds_karp: Vec<Edge<u32, OrderedFloat<f64>>> = self.edges_raw.iter().map(|((from, to), cap)| {
-            ((*from, *to), OrderedFloat(*cap))
-        }).collect();
-        let profiling_end = Instant::now();
-        info!("PROFILING Rust_PhillipsDessouky::max_flow scaling to OrderedFloat time: {:.10}s", PhillipsDessouky::profile_duration(profiling_start, profiling_end));
-
-        let profiling_start = Instant::now();
-        let (flows, _max_flow, _min_cut) = edmonds_karp::<_, _, _, SparseCapacity<_>>(
-            &self.node_ids,
-            &self.source_node_id,
-            &self.sink_node_id,
-            edges_edmonds_karp,
+    /// Find the min cut of the DAG annotated with lower/upper bound flow capacities.
+    ///
+    /// Assumptions:
+    ///     - The capacity DAG is in AOA form.
+    ///     - The capacity DAG has been annotated with `lb` and `ub` attributes on edges,
+    ///         representing the lower and upper bounds of the flow on the edge.
+    ///
+    /// Returns:
+    ///     A tuple of (s_set, t_set) where s_set is the set of nodes on the source side
+    ///     of the min cut and t_set is the set of nodes on the sink side of the min cut.
+    /// 
+    /// Panics if:
+    ///     - Any of the assumptions are not true.
+    ///     - No feasible flow exists.
+    fn find_min_cut(&mut self) -> (HashSet<u32>, HashSet<u32>) {
+        // In order to solve max flow on edges with both lower and upper bounds,
+        // we first need to convert it to another DAG that only has upper bounds.
+        let mut unbound_dag: LowtimeGraph = self.dag.clone();
```

Comment:
Please see if any of the `self.dag.clone()` calls in this method can be eliminated. This is probably a very expensive operation. Since this struct is constructed from scratch each iteration, in-place mutating stuff would be fine.

Reply:
There are 3 places we clone `self.dag`:
- initializing `unbound_dag`
- initializing `residual_graph`
- initializing `new_residual`

For initializing `unbound_dag`, I don't think we can remove the `clone` because the original dag is needed for construction of `residual_graph`.

For initializing `residual_graph`, I don't think we can remove the `clone` because we need the original dag edges in the construction of `residual_graph` itself. In particular:
```rust
        let mut residual_graph = self.dag.clone();
        for (u, v, _dag_edge) in self.dag.edges() {
            // Rounding small negative values to 0.0 avoids pathfinding::edmonds_karp
            // from entering unreachable code. Has no impact on correctness in test runs.
            let residual_uv_edge = residual_graph.get_edge_mut(*u, *v);
            let mut uv_capacity = OrderedFloat(residual_uv_edge.ub - residual_uv_edge.flow);
            if uv_capacity.into_inner().abs() < self.fp_error {
                uv_capacity = OrderedFloat(0.0);
            }
            residual_uv_edge.capacity = uv_capacity;

            let mut vu_capacity = OrderedFloat(residual_uv_edge.flow - residual_uv_edge.lb);
            if vu_capacity.into_inner().abs() < self.fp_error {
                vu_capacity = OrderedFloat(0.0);
            }

            match self.dag.has_edge(*v, *u) {
                true => residual_graph.get_edge_mut(*v, *u).capacity = vu_capacity,
                false => residual_graph.add_edge(*v, *u, LowtimeEdge::new(
                    vu_capacity,
                    0.0, // flow
                    0.0, // ub
                    0.0, // lb
                )),
            }
        }
```

For initializing `new_residual`, this one is more ambiguous but I don't think we can remove the `clone` because, while constructing `new_residual`, we have to loop over the original edges while modifying the `new_residual` edges:
```rust
        let mut new_residual = self.dag.clone();
        for (u, v, edge) in self.dag.edges() {
            new_residual.get_edge_mut(*u, *v).flow = edge.ub - edge.flow;
            new_residual.add_edge(*v, *u, LowtimeEdge::new(
                OrderedFloat(0.0), // capacity
                edge.flow - edge.lb, // flow
                0.0, // ub
                0.0, // lb
            ));
        }
```
If we added new edges while iterating over edges, I think this can lead to undefined behavior in rust. An alternative would be to `clone` just the edges, but the edges make up most of the memory footprint anyway so I don't think this would bring any noticeable performance boost. What do you think?

Reply:
Thanks for looking into this very thoroughly! Doing partial `clone`s won't be very noticeable as you said. Let's leave everything as is.